### ISSUE TITLE
API Concerns with Structured Clone for Wasm Modules

### ISSUE THREAD

**Issue by flagxor** [WebAssembly/design#972]
Some have expressed concerns that structured clone for wasm modules is a bad choice of contact point into the web platform.
In particular, it has these drawbacks:
* It encourages modules with the origin in which WebAssembly.compile was done, rather than the url of the bytes came from.
* It encourages use of IndexedDB as an application managed compilation cache, rather than leveraging the existing caching mechanisms in the browser.  Requiring middleware vendors to get the caching pattern right.
* It encourages use of postMessage for module for cross-origin shared modules.

We have casually assumed that the current mechanism would soon be augmented by support for WebAssembly.compile of a ReaderStream (for instance from fetch), to allow streaming compilation.
We've also discussed the possibility of surfacing wasm modules as es6 modules.

Have we found the right contact point?
Alternatives exist such as:
* Expand the Cache (service worker API) to support wasm compilation and explicit caching of wasm code, and knowing when it's actually compiled.
* Ensure Foreign fetch allows the possibility of sharing a wasm module cross-origin.
* Pass an URL instead of an ArrayBuffer to WebAssembly.compile. Using things like blob urls for dynamic cases.

We've already implicitly assumed we'll want to support more options.
Will any of those subsume the current interface to the point we'll want to deprecate it at some point?
Do we want to consider limiting the scope of structured cloning to plan for better future patterns?

**Comment by titzer** [comment]
Another option which I personally prefer would be to allow a new XmlHttpRequest response type (e.g. `WebAssembly.Module`), which doesn't introduce new API surface area and is an even stronger hint to the browser to start, e.g. streaming compilation while downloading.

**Comment by lukewagner** [comment]
The main two specific concerns that I've heard so far that make sense to me are:
* `window.postMessage` and `window.onmessage` are error prone APIs that are generally discouraged and we shouldn't be encouraging use of these APIs for people wanting cross-origin caching/reuse
* use of IDB to cache code in insecure contexts makes it easier for a one-time MITM attack (against the insecure origin) to permanently poison the site

If we all agree on these concerns, then I think both could be addressed conservatively by:
* disallowing structured clone of a Module to a window (only allowing postMessage to workers)
* disallowing the storing of a Module in IDB in insecure contexts (which follows from the [general](https://www.chromium.org/Home/chromium-security/prefer-secure-origins-for-powerful-new-features) [principle](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) of restricting powerful new features to secure contexts)

Looking forward, and building on basic structured cloning support, we could then incrementally evolve to:
* have WebAssembly.compile/instantiate take a stream, [as has been proposed](https://github.com/WebAssembly/design/blob/master/FutureFeatures.md#streaming-compilation)
* integrate with Service Worker and the Cache API, leveraging [Foreign-Fetch](https://github.com/w3c/ServiceWorker/blob/master/foreign_fetch_explainer.md) to support a safer-by-default way to cache/reuse code across origins

I think this strategy wins by supporting all the short-term caching and code-sharing needs of wasm without blocking on Service Workers (with new, to-be-defined APIs) in all browsers and also providing powerful primitives (viz., structured clone of modules) that can be used in other ways that wouldn't otherwise involve Service Workers ([EWM etc](https://github.com/extensibleweb/manifesto)).

There's also the idea to push all the caching into the browser (having all the APIs take URLs, using Blob-URLs for dynamically-generated content).  Just from a predictability point of view (and taking into account all the arguments for Service Workers over the HTTP cache), I think this is not a great option.  I totally agree we should make loading a wasm module as easy as a <script> tag, but I think that even in this case, caching should be handled by a Service Worker that intercepts the <script> request.

(cc'ing @annevk @bzbarsky @wanderview asutherland for any comments)

**Comment by bzbarsky** [comment]
> window.postMessage and window.onmessage are error prone APIs that are generally discouraged

This is the first I've heard about this so would appreciate concrete pointers.  In particular, postMessage is supposed to be _the_ API for cross-site communication, so if it's being dicouraged, what is being encouraged instead?

**Comment by lukewagner** [comment]
(Mostly forwarding @jeisinger's comments, so feel free to chime in)
@bzbarsky Maybe I overstated the case, but just these [Security Concerns](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#Security_concerns) about needing to verify the sender's origin.

**Comment by bzbarsky** [comment]
Well, right, you have to do that.  :)

**Comment by titzer** [comment]
Other advantages of using an XHR response type are:

- the browser gets to prepare for streaming compilation as soon as possible, e.g. even as it starts network requests
- WebAssembly.compile() is nominally implemented in the JS engine and it'd be weird for it to either have an API out to its embedder to handle URLs, or for it itself to be implemented in the JS engine
- XHRs inherently have origin URLs attached to them, whereas Streams do not necessarily.

**Comment by esprehn** [comment]
As discussed in the other thread, I don't think postMessage of wasm modules makes sense. It seems to be addressing the issue that `.compile()` doesn't accept a URL like JS/CSS/etc. so the underlying system can't implement streaming and caching for you like we do for other resources.

In general I don't see a difference between WASM, JS, CSS, Images, and the other resource types. They all are:
- Expensive to parse and compile.
- Used in multiple frames on the same page.
- Need async and stream processing.
- Need caching.

So far the platform has handled the resource caching implicitly. Having the platform handle it has been hugely beneficial to us as we've evolved the underlying system to use different caching strategies over the past several years. Implementing an effective strategy yourself in script is **very** difficult because you don't have access to things like the system memory limits, pressure signals, number of open tabs and frames, what CPU state the device is in, if it's plugged in, etc.

Separately given the security issues of sending compiled code across the origin boundary, I don't think we can allow sending pre-compiled code cross origin. Anything you send will need to be recompiled on the other side. It's basically the same as sending a string and asking the other end to call eval() on it in JS today.

I think what's missing here is a `.compile(url) => Promise<WebAssembly.Module>` API, integration with `<script src>`, `<link rel=preload>`, integrity hashes, etc.

@titzer Having .compile() take a URL which is implemented in the embedder is no more weird than ES modules having import statements that need to be fetched by the embedder.

**Comment by esprehn** [comment]
@drufball @bfgeek

**Comment by wllang** [comment]
@esprehn The performance of wasm code depends on baking in information that depends on the target features and the resources allocated or promised. Many uses case have been deferred to being implemented in user code in a translation stage between the data streaming in and the web browser wasm compiler. For example: user defined code compression strategies; dealing with differences between web browsers in their wasm implementations; taking advantage for performance enhancements that might have spotty deployment; etc. Perhaps this strategy of deferring these use cases was just wishful thinking and not practical, but it might have a lot of benefits if practical.

This translation stage might practically run in a separate context if that helps, a context with limited inputs so that it could be re-run with only those inputs to generate the same output, but it still needs input of the target features and some of the allocated or promised resources.

This translation stage might need to re-run even between instances on the client side, for example to deal with feature changes between browser versions, or even just changes in the available resources. E.g. The first instance of a game might get a lot or fast memory, and the second only a limited amount of slower memory, both needing separately compiled binary code to run.

**Comment by jeisinger** [comment]
I would like to see an API that allows for strongly tying a WebAssembly module to an origin. We should require CORS for exceptions throwing inside a cross origin module to bubble to the embedder, and we should be able to do mixed content checking.

I think that once we tie modules to origins, all use cases that currently require structural cloning would naturally work.

You wouldn't need the other origin postMsg'ing a module to you, you can just request it by specifying the URL. You also don't need to have a message handler with proper origin checking, the browser does it for you. You don't need to store the module in an IndexedDB, the browser can do that for you as well.  You can have a service worker that handles upgrading in the background. Once we have foreign fetch, you can support stuff like sharing an game engine on one origin with many other origins only hosting the actual game content.

My comment about message handler being suboptimal from a security point of view is that I don't that loading a cross origin script should require cross origin messages at all, and then not requiring that is better than requiring a message handler hoping that every site gets the origin checks right.

Also, if we add new response types, let's add them to the fetch api, not XHR :)

**Comment by jayphelps** [comment]
> Also, if we add new response types, let's add them to the fetch api, not XHR :)

fetch does not support cancellation, and since the latest proposal [was withdrawn](https://github.com/tc39/proposal-cancelable-promises/issues/70). I would much prefer that any response types supported by fetch were also supported by XHR so we can optionally cancel them.

**Comment by flagxor** [comment]
I believe using the collection of in-progress standards described here will make sense for the more common use cases once these APIs are finalized and available in all browsers. This was already the longer term direction we had intended to go, including rationalizing wasm modules in some way with ES6 modules.

However, I'm also skeptical that we can get the proper caching behavior right, in one go, in all the browsers in a fashion sufficiently coordinated to make the Web an appealing target for this kind of content.
A large part of the motivation for WebAssembly is to provide lower level mechanisms, because the high level mechanisms of the Web haven't consistently allowed developers to do what they need to be performant, even when they are highly motivated to do so (recall for this content the bar is identical content running as a native app).

Structured clone is definitely more low level than what we ideally want folks to use long term. But it is also expressed at a level that allows one to quickly determine that it exhaustively covers all the ways you might want to use it.

I have a hard time readily assessing what exactly the union of Service Workers, Streaming Fetch, Foreign Fetch, some kind of explicit/implicit code caching extension to the Cache API / XHR / Fetch, permits or denies developers. Can I cache content I generated synchronously? It also seems especially unclear what level of functionality this will make available in various browsers (as not all of them support the same set). 

It's also useful to keep in perspective that the majority of content initially targeting wasm will come thru emscripten or Unity, which prescribes JS boilerplate we can vet for generality.

Besides the case of allowing storage to indexeddb in insecure origins (which I'm fine with disallowing), I'm not sure a concrete security concern has been raised. That said, I believe limiting things to same origin + secure origins for indexeddb storage is a reasonable starting point, as folks seem very worried bad patterns will get baked in.

Failing that, I would hope those suggesting other API surfaces can clarify whether those surfaces have support in all the major browsers, and what the polyfill options are when they don't (i.e. what do browsers without cache api do?, most other use cases can be polyfilled with indexedb + XHR). I'd be more comfortable with the suggestion of other contact points with the web platform, if it was clear that contact point can work additively in most forthcoming browser versions with only the addition of webassembly.

**Comment by pizlonator** [comment]
@esprehn link to separate thread discussing passing modules in postMessage?

@lukewagner I think I agree with what you're proposing. Specific thoughts:

> The main two specific concerns that I've heard so far that make sense to me are:
> 
> window.postMessage and window.onmessage are error prone APIs that are generally discouraged and we shouldn't be encouraging use of these APIs for people wanting cross-origin caching/reuse
> use of IDB to cache code in insecure contexts makes it easier for a one-time MITM attack (against the insecure origin) to permanently poison the site
> If we all agree on these concerns, then I think both could be addressed conservatively by:
> 
> disallowing structured clone of a Module to a window (only allowing postMessage to workers)

Makes sense.

> disallowing the storing of a Module in IDB in insecure contexts (which follows from the general principle of restricting powerful new features to secure contexts)

I think I agree with this.

> Looking forward, and building on basic structured cloning support, we could then incrementally evolve to:
> 
> have WebAssembly.compile/instantiate take a stream, as has been proposed
> integrate with Service Worker and the Cache API, leveraging Foreign-Fetch to support a safer-by-default way to cache/reuse code across origins
> I think this strategy wins by supporting all the short-term caching and code-sharing needs of wasm without blocking on Service Workers (with new, to-be-defined APIs) in all browsers and also providing powerful primitives (viz., structured clone of modules) that can be used in other ways that wouldn't otherwise involve Service Workers (EWM etc).

I agree with not gating WebAssembly on other things. WebAssembly is a complex beast, so it shouldn't be too coupled to other complex beasts.

I think that if you enable WebAssembly in JSC then you will get it via the Objective-C API, too. That seems pretty neat, but in native apps there is no pre-existing caching engine to fall back to and it makes sense to let the client just do it however they like.

> 
> There's also the idea to push all the caching into the browser (having all the APIs take URLs, using Blob-URLs for dynamically-generated content). Just from a predictability point of view (and taking into account all the arguments for Service Workers over the HTTP cache), I think this is not a great option. I totally agree we should make loading a wasm module as easy as a <script> tag, but I think that even in this case, caching should be handled by a Service Worker that intercepts the <script> request.

Totally. Just as WebAssembly is engineered to get around all the wild dynamic JITing we do, it should probably get around the wild caching, too. Any claim otherwise has to answer: exactly how does the cache guarantee the same determinism as the current API?

@flagxor You said:

> Will any of those subsume the current interface to the point we'll want to deprecate it at some point?

I'm not convinced that any of them subsume the current interface, but that may just be that I'm missing some context.

If I want the compiling megapause to happen once in a multi-worker app, how do I do it using the current interface and the alternate interfaces?

**Comment by flagxor** [comment]
Agreed I'm not sure it does cover all the cases.

As for your multi worker use case:

For the current api you compile once in one worker or via the async method on the main thread, then post it around.

With the other, presumably you need to post around a message to let everyone know the cache has been prime. One worker would have to compile and added things to the cache (through the cache API which isn't available everywhere), and also check that compilation is done (through some as yet unspeced API).

I suppose if we imagined some clarification of how wasm is rationalized as an es6 import, importScripts in each worker could arrange for a single compile to be shared.

**Comment by pizlonator** [comment]
@flagxor Thanks for that explanation.

> For the current api you compile once in one worker or via the async method on the main thread, then post it around.

That is what I guessed.  This feels very natural.

> With the other, presumably you need to post around a message to let everyone know the cache has been prime. One worker would have to compile and added things to the cache (through the cache API which isn't available everywhere), and also check that compilation is done (through some as yet unspeced API).
> 
> I suppose if we imagined some clarification of how wasm is rationalized as an es6 import, importScripts in each worker could arrange for a single compile to be shared.

This feels less intuitive.

The compiling megapause is huge, and the result it produces could also be large - as big as an app binary. We want the client to be able to easily tell when the megapause happens and how long the result lives. Our current API gives us this, so we should stick with it.

Is anyone proposing any security-related extension to the current API beyond what @lukewagner proposed? Does everyone concur that @lukewagner's fix is sufficient?

Also, I don't understand how @titzer's XHR idea relates to this. Is this related to the security and API consistency concerns, or is this an unrelated optimization?

**Comment by wllang** [comment]
> For the current api you compile once in one worker or via the async method on the main thread, then post it around.

Except that this will not work in general without re-compilation, and if a user translator is in that pipeline than it would need to be part of the module posted plus the original input data.

Also the compiled code will be specific to allocated or promised resources and it does not appear that the sender JS code can practically know the details of the resources that the receiver will have? So the strategy of posting a compiled module from one context to another has rather limited applicability.

What would seem practical is for a number of threads to share the same compiled code as they could all be working with the same resources, but perhaps this could be a `fork` style operation and a separate matter.

I repeat my suggestion that a pipeline be defined that includes the user translator and the inputs which will include the allocated or promised resources, so that the compile code can be generated from these inputs. The web browser would then be responsible for managing the caching and compilation based on these inputs. Some of the inputs might be URLs, and others internal details of the implementation that get baked into the compiled code.

This also requires resource management, to obtain the promised resources. In particular the promised linear memory will be needed and some characteristics of this will need to be exposed to the user translator, such as the size, but other details might be internal. If the promised linear memory ends up not being fulfilled then it might even be necessary to re-try this process.

For features this might mean that the model of generating and attempting to validate code to do feature detection is not a good fit, rather the features requested or used might need to be inputs exposed to the user translator and used as keys in the cache. The cache keys would be best no more specific than the features and specific inputs used.

**Comment by annevk** [comment]
I was asked to chime in by a couple of folks and only have a basic understanding of the setup, so please bear with me.

Could someone explain what the message passing of wasm code enables? Is the purpose here to compile the code once and then execute the compiled code in multiple workers? Meaning that the compiled code would effectively be shared (readonly I hope)?

And the angst is because this is introducing a new primitive that doesn't exist for (JavaScript) code on the web today?

More minor things: 1) @jeisinger, if you have an object you can pass around, there's no need for origins. Message passing effectively has CORS builtin. 2) XMLHttpRequest won't see any extension. Any extensions will be to `fetch()`. 3) Is there a story for wasm and `<script type=module>`/`new Worker(url, { type: "module" })`?

**Comment by wllang** [comment]
@annevk I'll let other people explain their question, but I understand they do want to use message passing and the IndexedDB to enable management of the compilation and caching in JS.

I don't think their approach will work well, but could I explore an alternative as a separate thread of discussion rather than having an absolute position, and so ask how might `fetch()` work with the technical fact that for best performance the code needs to bake in code specific to the resources at hand which might depend on how the linear memory is allocated? Could `fetch` support a user defined translation stage between the content streaming in and the compilation of that content, and could it support inputs into this pipeline that depend on resources allocated or promised?

I am not sure how valid the analogy is, but it might be similar to an image resource being fetched for which part of the decoder is user defined code (which might decode from a compressed format to a simpler bit map image format) and this decoder might need to know the target device pixel encoding as an input and this might change depending on the resources a tab has and the target dpi etc which might change if the user scales the display or moves a window to another monitor or casts it to another device etc. How would you recommend abstracting this?

**Comment by annevk** [comment]
@wllang `fetch()` returns a promise for a response that settles once all the headers are in (but the body is not). I would recommend passing the body (a stream) of that response to your custom decoder that does the translation.

**Comment by lukewagner** [comment]
@annevk Sorry for lack of context; here's some:  So we have a JS object introduced by wasm, `WebAssembly.Module`, that we've currently defined/implemented to be structured cloneable because it is stateless (so not a JS function object, more like an unevaluated function literal).  A `Module` abstractly represents "code I've already compiled" (at some considerable expense) and so intuitively `postMessage`ing it to workers is a way to share already-compiled code with workers and storing a `Module` in IDB is a way to serialize the compiled code.  (It's up to the browser to implement these optimizations, of course.)

**Comment by annevk** [comment]
Given the concerns raised so far with that approach I'm sympathetic to @esprehn's concerns. It does indeed seem rather novel and not much existing infrastructure is being reused. The deduplication offered with `<script type=module>` could be used here potentially, although for workers you would need more since you want to deduplicate compilation and code across workers, whereas the module map is currently global-scoped. Supporting something like `new Worker(..., {type:"wasm"})` doesn't seem out of the question however. Has that been explored?

(The security concern with `postMessage()` seems arbitrary. If you allow non-trusted `ArrayBuffer`s or `Module`s to be posted, what is to say you won't allow non-trusted URLs (but nevertheless secure, since they are attacker-controlled) to be posted?)

**Comment by lukewagner** [comment]
@annevk `postMessage` and IndexedDB are existing infrastructure that are being reused, and they're also in all browsers :)  The question isn't whether we should integrate with `<script type=module>` and other ways to load ES6 modules -- I think we should -- the question is whether there are any hard reason to *stop* supporting structured clone of `Module` and *only* allow sharing via URL (and effectively blocking predictable wasm code sharing/caching on SW + new SW APIs).  So far, I don't see any new hard reasons.

**Comment by pizlonator** [comment]
I completely agree with @lukewagner's point. Module is a thing that can be passed around with postMessage and doing so has clear semantics, so forbidding it would be silly.

@annevk 

> The deduplication offered with <script type=module> could be used here potentially, although for workers you would need more since you want to deduplicate compilation and code across workers, whereas the module map is currently global-scoped. Supporting something like new Worker(..., {type:"wasm"}) doesn't seem out of the question however. Has that been explored?

WebAssembly can be used for simple things like "load code from URL and run code" but it exposes something more powerful: you can programmatically generate WebAssembly code in JS or in WebAssembly, call `instantiate` or whatever to compile it, and then you can run the thing you compiled. In that world, the user want to think of the code that was generated as just an object - not a named resource.

**Comment by annevk** [comment]
> postMessage and IndexedDB are existing infrastructure that are being reused, and they're also in all browsers :)

The way you are using them is not a way that any other new resource type is using them. Not reusing existing fetch infrastructure is also rather novel. I'm not sure ignoring @esprehn's post is helping.

> In that world, the user want to think of the code that was generated as just an object - not a named resource.

Yeah, understood. Mostly trying to mediate at the moment. (I do think that if we were to offer such functionality we should also have it for JavaScript. Not sure why it would have to be specific to WebAssembly. Compilation is a bottleneck for JavaScript too.)

**Comment by flagxor** [comment]
> The way you are using them is not a way that any other new resource type is using them. Not reusing existing fetch infrastructure is also rather novel. I'm not sure ignoring @esprehn's post is helping.

JS and other resources with large decode costs would also benefit from explicit caching / passing of compiled code in a small subset of use cases. This probably has not happened (yet) because JS also suffers from having a context dependent meaning, and because the majority of these resources are small.

The normal case for JS is that a script tag just works for most things. Adding a basic service worker makes offline work. By comparison, with 50MB modules being common, even a "normal" Wasm module confronts what would be the advanced case for JS, i.e. the need to show download progress, background the compile, and ideally manage incremental updates and sharing more carefully.

Providing direct low-level mechanisms (as long as they're secure) is a good way to avoid baking-out use cases.
WebAssembly for example intentionally avoided describing its dynamic linking in terms of ES6 modules, instead providing a lower level mechanism, so as not to exclude things that C/posix style dynamic linking allows. In fact, we picked a mechanism even lower level than C, so that non-C dynamic binding wouldn't be ruled out.
Similarly SharedArrayBuffers avoided only providing high level synchronization primitives like mutex/condition vars/queues, as futex + atomics are even more expressive, and because the others can be implemented in terms of them.

As we start to see small chunks of Wasm code in JS frameworks, we'll surely want the implicit mechanisms to also be present for good interoperability. But I'm unconvinced it's wise to rule out more general mechanisms, as Wasm is explicitly intended to fill in the 10% case not well covered by JS currently.

**Comment by pizlonator** [comment]
@annevk 

First of all, @esprehn's post cites another discussion that he hasn't provided a link to. I'd like him to either describe that discussion or show a link. Otherwise, it's hard to follow.

You said:

> Yeah, understood. Mostly trying to mediate at the moment. (I do think that if we were to offer such functionality we should also have it for JavaScript. Not sure why it would have to be specific to WebAssembly. Compilation is a bottleneck for JavaScript too.)

In JS, the compilation happens whenever the VM feels like doing it because the compilation is coupled to the dynamic type inference, which requires a profiling stage. We're not going to be providing a compiling API for JS anytime soon, because that would mean locking in the timing of when it happens and we have found empirically that we want this to be tunable in the VM. For example, if we slow down our compiler by adding a new kind of optimization then we would want to change when the compiler runs to offset the increased cost.

On the other hand, WebAssembly compilation is a well-understood procedure. We can specify compilation for WebAssembly because the entire WebAssembly language was designed from the ground up in order to enable us to define what it means to compile it. No such thing exists in JS.

Therefore, I don't think it's useful to assume that the way JS gets compiled should have anything at all to do with how WebAssembly is compiled.

**Comment by wllang** [comment]
@annevk So `fetch` returns a stream for the body of the deployed blob, and I presume manages the caching of that blob. However the use case being discussed is far more general, and includes the caching and reuse of the compiled code generated in part from that blob, and just having a stream does not solve the use case that posting modules and storing them in the indexdb is intended to address.

I would also note that people keep characterising what they are proposing as "predictable wasm code sharing/caching on SW + new SW API", however this does not fit the technical realities. The fact is that two separate contexts might allocate linear memories with different characteristics which require different compiled code and the sender will not know this so that approach will just not work without more design work.

For example, an implementation could restrict all wasm linear memories to having the same capabilities so either all run in separate processes or all use particular sized guard zones or all use inline bounds checks. Flags might be added to wasm compilation that hints at a preference to optimize for one of the cases and that might be used as a cache key, etc.

For example, it might be possible abstract these contexts so the requesting context might first obtain a promise for a linear memory (a somewhat opaque object that is a descriptor of the linear memory) and post it to the sender and the sender might then compile the wasm module for that specific descriptor (looking up it's indexdb cache and using that descriptor as a key) and then send a speculatively compiled module back, and retry if needed.

People are not giving you all these details of such a plan and there might be very significant constraints.

Some other good points were made, such as wanting to support the use case of code generated without a pipeline that is based on a URL. Some of these usescases would be compiling in an already running wasm context with already allocated resources so might well be able to obtain a descriptor of this as an input to compilation and a cache. Some of these use cases might be able to build a provenance descriptor that allows the input to be regenerated, and in other case the only practical cache strategy for these might be to use the dynamically generated input as the cache key.

There are lots of precedents outside the JS sphere, for example scientific computing architectures do a lot of caching and manage large resources, so it's not new computer science. These are also not recent issues for wasm and were raised long ago such as https://github.com/WebAssembly/design/issues/378 and https://github.com/WebAssembly/design/issues/376

I just don't think people have pushed the design far enough to be asking for the merits to be judged, it's a WIP, but it would be good to have more people pondering a good path.

**Comment by lukewagner** [comment]
@annevk The basic symmetry argument seems reasonable but I think it shouldn't prevent us from starting at a separate point from JS and incrementally moving both wasm and JS towards each other.

So on the wasm side that means the things we've mentioned above and were already planning to do (that's the nature of an MVP; all sorts of good stuff is left out in the first iteration).

And, actually, on the JS side, while I agree with @pizlonator that we're not going to have much success caching/sharing fully-optimized machine code, I think a new, opaque, stateless, structured-cloneable uninstantiated JS module object could potentially have major perf benefits and be worth considering in the future.  At the very least, an engine could store the chars in an ideal format (the ideal encoding + compression, noting function boundaries and maybe upvar aliasing, and noting the absence of early syntax errors) so that there was zero up-front scanning/syntax-error-only-parsing required when instantiating that JS module.

But I don't mean to get off on a JS-load-time tangent: I just mean to make the point that this isn't the end of the story and we should expect continued evolution based on our experience shipping v.1.  What's important is that we have something that works now, with relatively predictable performance, multiple implementations, and an evolution story.  I'd like to avoid increasing our scope if it's not absolutely necessary.

**Comment by annevk** [comment]
Thank you all for the detailed replies. I agree that we need to hear more from @esprehn and @jeisinger or get a pointer to the relevant thread.

**Comment by wllang** [comment]
@annevk I recalled some other goals that appealed to me personally in past discussions. Let's say that there could be a description of what resources were needed to launch a wasm app, so a description of the URLs to fetch and the pipeline to process them and the linear memory that was needed and the stage at which it needed to be allocated. Then the web browser would have some very useful information to schedule that computation, and to take advantage of a caching. Plus some very useful information to make decisions for the user such as being able to inform the user that they need to close some tabs to free memory to be able to run a demanding app well before it just fails to load, or inform users that they could only run the app in a degraded mode without closing tabs, or for a demanding game perhaps to go all in and allocate everything for that app. The solution being discussed above, of having some user JS mange all this, would not appear to practically support such good outcomes, it could not practically know other conflicting resource usage by the web browser and could not control the scheduling and execution of the pipeline. I do appreciate the goal of exposing as much as possible to developer customization, but perhaps there is some other light weight solution with more potential, and perhaps you can see potential outside just the wasm pipeline.

**Comment by annevk** [comment]
So I understood from @jeisinger that the referenced thread is a) internal, b) doesn't contain much information, and c) that loading from a URL is very much a requirement.

I'd like to understand how with URLs these use cases are addressed:

1. Having multiple workers run the same in-memory instance of the compiled code. (As a bonus, do so with only a single fetch, which seems like a requirement given the size of the resources involved.)
2. Manipulate the resource before compilation and then cache the compiled result. (As a bonus, do so without service workers to not block on Edge and Safari.)

It would also be good to know why Indexed DB is bad for this, given that the Cache API is equivalent for all intents and purposes, except that it cannot store objects.

**Comment by pizlonator** [comment]
Also, I'd like to know why loading from a URL is a requirement. 

-Filip

**Comment by esprehn** [comment]
When I said other thread I meant this discussion:
https://github.com/WebAssembly/design/issues/956#issuecomment-272563878

I wasn't referencing any internal discussion.

URLs can be de-duplicated in the fetch the same way as web sites that start multiple JS workers, for example image encoding libraries will start 4 workers to do work in parallel all loading the same script. That's a single fetch and compile inside browsers. To share the same memory space you use SharedArrayBuffers sent between the workers.

Note also that compiling from a byte array is equivalent to eval() and will be blocked by CSP policy. So loading from a URL is a requirement here.

In general wasm seems to be in the "inline script only" situation right now. We need the "external script" facility and to integrate it with the rest of the machinery for integrity hashing, <link rel=preload>, and all the other goodness we're building for various resource types.

**Comment by wllang** [comment]
@annevk

> 1. Having multiple workers run the same in-memory instance of the compiled code. (As a bonus, do so with only a single fetch, which seems like a requirement given the size of the resources involved.)

This is not going to be possible in general. For example the x86 32-bit code will likely want to bake in the linear memory base address into the offset field on memory access instructions so unless they all share the same linear memory they can not share "the same in-memory instance of the compiled code". There are other reasons too such as differences in the sandboxing of the linear memory accesses. But this could work for some targets under some restrictions that an implement might be able to live with, but other implementations might have a higher performance bar.

Couldn't the 'single fetch' case be handled simply by the caching network layer, so it sees one request come in first and then the others and could schedule just one download and share it, so this seems ok assuming a network level cache can be leveraged?

The challenge might be a 'single translation and compilation' case, and that seems to be where the indexdb/clone solution is focused. So I guess there would be a context dedicated to the translation and compilation and caching of this code, and any other context needing this would put in a request so that it could look up the cache and order the translation and compilation. It would need some more design work to handle the opaque differences between the requesting contexts. But there would still be no coordination between these contexts, so for example a number of contexts might schedule a translation and compilation pipeline for different resources at the same time hitting web browser resource limits.

One obvious solution (well known solution for resource planning problems) is to have a description that can be used to globally plan these pipelines. There are many ways this description would be built and stored which might be somewhat orthogonal to there being a description. The description might include that a URL needs to be downloaded, that a linear memory needs to be allocated or will be used, perhaps other features to be used, basically any input that affects the output, and that a translation and compilation pipeline needs to be run. Then if four workers each submit the same description then the web browser can globally schedule this computation, and if there is some common shared stages then it can share only those and schedule the others as needed based on resource planning. For example the four workers share the same URL in this example which might be cached or can be download just once, but they might all need to separately translate and compiled the code, or some might share the translation stage and just need separate compilation, etc. I don't see how this can be implemented in user level JS code as it is just not practical to give it access to such global web browser information and control.

The descriptions might be in an attribute to the script element, or might be built by JS in a number of calls, or JSON, etc. Could this integrate with promises somehow, to have a chain of promises that produce an output and that could form a description of the work to be done. Perhaps someone up to speed with the cutting edge web standards has some insights, or should we just put together something that works for wasm for now, something very simply that can be trivially subsumed by something more general in future?

**Comment by pizlonator** [comment]
@esprehn 

> When I said other thread I meant this discussion:
#956 (comment)

I think there's an easy answer to your question: it's not a goal of WebAssembly to have compiling be cached using the browser's other mechanisms.  Instead, it's a goal to let the client manage this using whatever mechanism they like.  We made that the goal because we confirmed that this is what users would prefer.

In this thread, we have already offered many reasons why our current API is preferable over what you propose.  Can you provide a rebuttal to the points already made?

> URLs can be de-duplicated in the fetch the same way as web sites that start multiple JS workers, for example image encoding libraries will start 4 workers to do work in parallel all loading the same script. That's a single fetch and compile inside browsers. To share the same memory space you use SharedArrayBuffers sent between the workers.

Passing around URLs to reference an object when you could pass around an object is a worse API for our use cases.

We have a choice whether to represent data as a named resource or whether to represent it as an object. If it's an object then its lifetime is determined by the GC or by wherever that object is cloned into (IDB and postMessage effectively have their own lifetime rules). On the other hand, a named resource needs to either stay around forever or needs to be able to completely describe the resource used as an input to compute it, so it could be recomputed.

It's cumbersome and pointless to regress the WebAssembly design from using objects to using names, since WebAssembly's current use cases benefit from GC-style lifetimes.  This is what people are asking for, not the thing you suggest.

> Note also that compiling from a byte array is equivalent to eval() and will be blocked by CSP policy. So loading from a URL is a requirement here.

One of the dangers with eval is that you can pass a string to it.  A string from your comments section might now become code if you're a bad programmer.

You can't pass a string to WebAssembly.instantiate.  So, if you're a bad programmer, the string from your comments section would cause instantiate to throw an error, because it wants a byte array instead.

On the other hand, if WebAssembly.instantiate took a URL and that URL was a string, and you're a bad programmer, then a string from your comments section might become a URL to instantiate.

> In general wasm seems to be in the "inline script only" situation right now.

I don't know what about WebAssembly is "inline" or "script".

The input to WebAssembly compilation is a byte array.  That byte array is an object.  It can come from anywhere.

> We need the "external script" facility and to integrate it with the rest of the machinery for integrity hashing, <link rel=preload>, and all the other goodness we're building for various resource types.

I have no idea what you just said.

**Comment by Kardax** [comment]
@pizlonator Loading from an `ArrayBuffer` is on the same security plane as `eval` because there is no check on whether the input is from a trustworthy source.  WASM's strictness (particularly the prohibition of multiple code sections) provides a high degree of protection from an injection attack, but this is not enough to completely eliminate risk.

**Comment by pizlonator** [comment]
@Kardax please describe the injection attack.

**Comment by Kardax** [comment]
@pizlonator A lot of code between the `fetch` response and `WebAssembly.instantiate` that provides a way for an attacker to modify the `ArrayBuffer` via URL input.  The combination of carelessness, bad design, and luck to achieve this vulnerability is much higher than with JavaScript, but as long as there's a gap between acquiring the `ArrayBuffer` and instantiating the WASM, there's room for stupidity.

**Comment by wllang** [comment]
@pizlonator 

> In this thread, we have already offered many reasons why our current API is preferable over what you propose. Can you provide a rebuttal to the points already made?

Why don't you take me on if you're so confident in the technical merits of your approach?

> On the other hand, a named resource needs to either stay around forever or needs to be able to completely describe the resource used as an input to compute it, so it could be recomputed.

An object which is a compiled wasm module does not solve the use case because it might need to be re-compiled when re-used? So a description that allows it to be re-compiled is one solution, and you have not put forward another but I can see it might be possible to pass around an opaque object describing the target context and to use that as a key for compilation and the cache.

A URL might just reference a description of how to compile the resource, it might not even be the wasm file, and that description might be small enough to be inlined too.

> It's cumbersome and pointless to regress the WebAssembly design from using objects to using names, since WebAssembly's current use cases benefit from GC-style lifetimes. This is what people are asking for, not the thing you suggest.

These are very loaded comments. It's not even clear that other people would support you on this - I don't think this is about 'objects' versus names rather an attempt to move complexity out of the web browser to the JS user code which sounds great but it seems to have some show stopper problems. 

We need a work environment that can yield design solutions not personal wins by jumping on people. I'd like to see some more ideas and would welcome the proponents of the clone approach extending it to address some of the problems raised, to show the future extensions that are expected to address these problems?

> please describe the injection attack.

Is there a W3C group working on CSP etc and might they have something to say and be able to articulate the issues very clearly.

Here's another use case challenge: run a wasm app with JS disabled? Avoid all the JS attack surface, rather just depend on the wasm sandbox. Having worked on the web browser JS implementations my faith in their security is rather thin!

**Comment by pizlonator** [comment]
@Kardax 

> A lot of code between the fetch response and WebAssembly.instantiate that provides a way for an attacker to modify the ArrayBuffer via URL input.

I agree that the possibility of a hole exists.

> The combination of carelessness, bad design, and luck to achieve this vulnerability is much higher than with JavaScript, but as long as there's a gap between acquiring the ArrayBuffer and instantiating the WASM, there's room for stupidity.

I think that the design goal of WebAssembly has always to be to allow some gap, so that the client can do essentially macro processing.

Let's try to be thoughtful of the trade-offs.  I'm going to try to describe them first without passing judgement.  Please chime in if you agree with my characterization.

Pros of current API:

- It has already been bikeshedded by WebAssembly stakeholders.
- We know that it satisfies the goals of WebAssembly, because we know that it works with all of the use-cases we have been studying. This includes use cases that do caching, workers, etc.
- Doesn't gate WebAssembly on anything that doesn't have universal browser support.

Cons of current API:

- It might have to be disallowed under CSP if at some time we conclude that the attack vector described by @Kardax is real enough. My understanding is that no such conclusion has been reached, only that it was proposed by @esprehn.
- Does not leverage smarts in the browser's cache.
- Does not leverage smarts provided by things (that aren't yet universally supported).

Some thoughts:

Even if CSP did forbid the current `instantiate` API, we would still want that API for standalone JS systems and we would still want that API for non-CSP environments.

This API does not preclude the browser from doing caching.

This API does not preclude leveraging smarts from other things.

**Comment by jeisinger** [comment]
Note that Chrome's implementation actually disables WebAssembly in the presence of CSP (can be whitelisted via unsafe-eval).

Please take the fact that we're having this discussion now as a sign that the structural cloning part has not been successfully bikeshedded with the relevant stakeholders. In fact, Chrome decided not to ship structural cloning of WebAssembly modules for now.

**Comment by jfbastien** [comment]
@jeisinger 
> In fact, Chrome decided not to ship structural cloning of WebAssembly modules for now.

Which part? `postMessage` to worker, IndexDB, and / or something else?

**Comment by jeisinger** [comment]
I'm interpreting your question as "did you implement any exceptions that would work even though WebAssembly modules don't support structural clone", is that what you meant?

The answer would be that no, there are no exceptions.

**Comment by jfbastien** [comment]
@jeisinger I meant:

* What does `postMessage` of a `WebAssembly.Module` do?
* What's the result of putting a `WebAssembly.Module` in IndexDB?

Same for other structured cloning.

IIUC you're saying: none of these work in Chrome 58. Is that correct?

**Comment by jeisinger** [comment]
correct (that is, not without command line flags)

**Comment by wllang** [comment]
@pizlonator 

> We know that it satisfies the goals of WebAssembly, because we know that it works with all of the use-cases we have been studying. This includes use cases that do caching, workers, etc.

I've raised the problems above, and no-one has refuted them, so this is an odd statement. If you can't refute them then I suggest that people can not accept this claim.

> Doesn't gate WebAssembly on anything that doesn't have universal browser support.

I don't buy this either, wasm is new and a web browser supporting wasm could reasonably be expected to support features it depends on. A web browser with wasm support and without support for posting module objects and storing them in an indexdb would not work either.

> This API does not preclude the browser from doing caching.
> This API does not preclude leveraging smarts from other things.

Right, but the web community will invest in it and if it is a dead end then that's a loss for the community.

Here's one alternative, a description of the pipeline. From this the web browser can schedule fetching the translator and allocating it memory and schedule fetching the source and allocating the linear memory. It could isolate the translator to prevent attacks on the pipeline and to ensure the output depends only on the defined inputs. The translator might even be updated to take advantage of new web browser features. It could be referenced from a script element in html by reference to this description resource. Is this something that people could refine or morph into something that would fit the web?

```
URL: https://example.com/game.wdes

Source: https://somecdn.com/game.enc
Source hash: 295f19....
Source size: 80M
Translator: https://anothercdn.org/decoder.wasm
Translator hash: ....
Translator memory: 16M
Memory tiers: 256M, 512M
Required features: threads, ...
Usable features: SIMD, ...
```

**Comment by annevk** [comment]
> That's a single fetch and compile inside browsers.

Is that actually true? We don't define such deduplication in Fetch. Servers could return unique resources each time. And I'm pretty sure there are some cases where such deduplication would break things. If that is a thing you think should work (or does already), please open an issue at https://github.com/whatwg/fetch/issues/new.

As for resource integrity and preloading, those can be used for resources acquired through `fetch()` too. But I guess the argument is that if you know it's wasm, you can compile the incoming stream (hoping the integrity ends up matching)? Requiring CSP unsafe-eval seems totally reasonable. It doesn't seem that much different from `eval()` indeed.

**Comment by lukewagner** [comment]
> In fact, Chrome decided not to ship structural cloning of WebAssembly modules for now.

This is surprising and frustrating given that, until recently, experimental wasm support in Chrome *has* included structured clone to the degree that various middleware vendors have already tested interoperable caching between Chrome and Firefox (with the expected massive improvement over a cold load).  Many of these same apps are already caching assets in IDB so caching the wasm alongside is natural.

At least in FF, we haven't seen any compelling security reasons to disable structured cloning and do not wish to regress users so will not be disabling structured clone (unless everyone else agrees to remove it from the spec).  Thus, tools can just `try { store.put(wasm) } catch(e) {}` and then Chrome will just end up never successfully caching.  I think this will (further) hurt the impression of overall (cross-browser) wasm load time, though.

**Comment by jeisinger** [comment]
I understand that this is disappointing, but I think it's important to stop treating this discussion as a security concern that might or might not exist, but as a discussion about an API surface that we'd be able to ship from a web platform perspective.

**Comment by lukewagner** [comment]
@jeisinger I totally get that SW is the shiny future here and could ultimately offer ideal performance and ergonomics for a majority (but not all!) use cases.  But it feels like there is an implicit assumption that using structured clone to persist and share things is being considered deprecated as a part of the platform.  If they are, that's one conversation.  But if they're not, then how is structured clone not purely additive to a platform that already has these concepts?  I know it's code, but so is a string of JS chars or a Bob of wasm bytes.

Is the underlying worry that adding structured clone will take away the motivation to follow through with future work (caching and SW additions) and thus you're trying to preserve a sense of urgency?

**Comment by wllang** [comment]
@annevk 

> We don't define such deduplication in Fetch. Servers could return unique resources each time. And I'm pretty sure there are some cases where such deduplication would break things.

So what happens if a html pages uses an images many times. Would web browsers have some sensible strategy of speculating that it was the same on each request and delaying launching multiple requests until at least the response headers were received.

> please open an issue at 

Is that a hint that this is going nowhere and that you need something far more concrete to consider?

> But I guess the argument is that if you know it's wasm, you can compile the incoming stream (hoping the integrity ends up matching)? 

Great point. Would the web browser know if it is a dynamic response after receiving the response headers, and couldn't it use a HEAD method to probe this first. Would that be enough to have good confidence to start streaming translation/compilation. Being able to version the resources seems a requirement for cache management.

> Requiring CSP unsafe-eval seems totally reasonable. It doesn't seem that much different from eval() indeed.

That would exclude some potentially great use cases for wasm in security critical web software. Wasm uses a much simpler compiler than JS, and offloads a lot of this to the producer, and implements a simple sandbox, so it might have been a great platform for running secure code. It is also expected to run some really big blocks of code and have a small set of operators, so the test coverage might be much higher. Could you ponder a solution that allows wasm to run in security sensitive contexts. Obviously the wasm JS API could not accept a buffer of code to compile without unsafe-eval, but can we still have a good solution that is secure.

@lukewagner 

> Is the underlying worry that adding structured clone will take away the motivation to follow through with future work (caching and SW additions) and thus you're trying to preserve a sense of urgency?

If code is deployed using this then that will create a legacy problem, and that's not a good start. It would be hard if not impossible to fill for a clone/indexdb custom solution. However if you could take a look at the proposal to use a description of the pipeline and adapt that to meet the use case you might be representing then this might be something trivial to fill in future.

**Comment by slightlyoff** [comment]
Hey all,

Late to the thread. Apologies for making it longer.

Speaking from the perspective of the web platform and loading in browsers, it seems important to note a few things:

 - Using `Response` and bodies _does not require Service Workers_. Doesn't even require the `Cache` API. You can continue to use IDB for code storage if you like.
 - Having a separate (more manual) loading path for your feature is a huge red flag. I'm going to file an @w3ctag issue on this to get review started. As a part of that, I'd like to better understand the use-cases for the fully manual loading and cloning. 
 - The notion of moving all WASM code into memory is...sub-optimal. One of the best reasons to re-use the `fetch()` types is the ability to load and store resources without extra copies and the ability to transparently take advantage of streams.

Those of us who feel the weight of web platform consistency really do want to understand the arguments for fully-custom behavior better and try to come up with some middle ground that meets your needs without creating parallel (WASM-only) behavior and types. For instance, there are proposals brewing for code optimisation hints for JS as well. Unifying them should be treated by everyone as a goal. If it's not possible, so be it, but we should evaluate the needs and work towards consistency.

Regards

**Comment by wllang** [comment]
@slightlyoff Thank you, some more input would be very welcome. Speaking for my own understand of the challenges:

* The resources (the encoded code) can be very large and it is likely that it is all needed before the web app can start, in contrast to small image resources or a streaming videos. Perhaps some new strategies are needed due to this difference.

* Personally I see great value in having a user defined decompression/translation stage in the pipeline. This could allow web developers to explore their own encodings independent of the web browser vendors and their agreement. It would also allow web developers to deal with differences between web browsers and to take advantage of enhancements that have spotty deployment etc. We could live without one wasm standard if there were such a layer that worked well, so this might be a critical component, and it might set the stage for web browsers innovating for years to come. There are a lot of enhancements planned for wasm.

* The output of a decompression/translation stage might be many times larger than the source, a significant blob to cache. Web browsers may well want to not even store it rather just generate it in a stream and regenerate it as needed. It might not be practical to keep this around as an input key to the compiled code cache.

* The code needs to be compiled before running in general. Some approaches might work around this somewhat, such as starting with an interpreter etc, but there is a practical use case for being able to compile it as it streams in and with the translation stage it is compiling the output of that stage. The compiled code could also be a significant sized blob.

* There are use cases in which the same code can be re-used. Some might use exactly the same code, and some might use it in different contexts that need a new translation and compilation pipeline to run.

* Given the above, I believe that practically some resource management is needed. One approach is to allow manual management of this in custom code on the client. So the manual code would be responsible for managing the upstream versions and re-compiling as necessary and would ultimately be passing around an object representing this compiled code, and hence the structured-clone/indexdb proposal.

* Personally I see technical problems with the 'manual' structured-clone/indexdb proposal approach, that the compiled code will depend on the target context so may well need to be recompiled when reused. Further the management of the resources such as CPU cycles and memory can be better managed by the web browser which has more global knowledge of these than it would appear practical to give to manual user code running in a web origin context. Compiling a large blob of wasm could tax the resource of a device, it may well need to schedule it, and it might help to be able to compile before allocating the linear memory (but the characteristics of the linear memory may be an input to the compilation stage).

* The wasm code runs in a new sandbox, and could potentially run without JS, or with a barrier between the two. It could even have API's that allow it to generate new code to be compiled and linked back into itself without exposing that potential to the larger JS code. There appear to be some compelling use cases for secure code here, and if the compilation were managed by manual JS then this separation might not be possible (unless the wasm code could have a restricted API for this all).

Personally I think borrowing some ideas from the scientific computing would help, so build a description of the tasks that need to be done and the resources needed so that the web browser can plan their loading and execution and optimize it and manage the cache. If getting this integrated into the web is going to take a big effort then wasm might use a somewhat throw-away description for now that might be easily subsumed by something much more comprehensive in future.

**Comment by wllang** [comment]
@slightlyoff There were no responses to my input about making the case for provisioning for recompiling code for the specific context, but here is some quick technical support for this position. These are run times for the zlib benchmark (a compression algorithm) running in wasm code. You can see that even my quick attempts at this using a 'pointer-masking' strategy and other approaches that might be specific to the context can give some useful performance improvements over that achieved so far by the web browser vendors. I presume they will want to catch up and it seems prudent to set the stage for this in coming years.

gcc x86 32-bit: 9.4 sec
custom pointer-masking wip x86 32-bit: 9.8 sec
firefox wasm x86-32: 15.3
v8 x86-32: 15.3

**Comment by jfbastien** [comment]
TL;DR: I think this will be addressed by #1048.
